### PR TITLE
Fix dotr -tasty

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: project/Build.scala [hunk 1]
```diff
@@ -504,15 +504,17 @@ object Build {
...
-        val _  = packageAll.value
+        val _ = packageAll.value
         val scalaLib = attList
           .map(_.data.getAbsolutePath)
           .find(_.contains("scala-library"))
           .toList.mkString(":")
 
-        if (java == "")
+        if (args.isEmpty) {
+          println("Couldn't run `dotr` without args. Use `repl` to run the repl or add args to run the dotty application")
```
Comment (smarter) [inline project/Build.scala]: Why not call the repl task at this point instead?

Comment (nicolasstucki) [reply project/Build.scala]: I tried with `repl.evaluated` but then `dotr` only ran `repl` ignoring the nonempty args. Then I just decided to show this message as we don't really need more in the sbt project.

Comment (smarter) [reply project/Build.scala]: sbt task evaluation is tricky, you probably need to use a dynamic task to get that to work.

Comment (nicolasstucki) [reply project/Build.scala]: That is what I thought, hence settled for the message. My main concerns were the factorisation of code and checking the it was executing correctly.

File: project/Build.scala [hunk 2]
```diff
@@ -640,6 +617,19 @@ object Build {
       }
   )
 
+  def dotDynTask(main: String) = Def.inputTaskDyn {
```
Comment (allanrenucci) [inline project/Build.scala]: I think `runCompilerMain` is a better name

File: project/Build.scala [hunk 3]
```diff
@@ -504,49 +504,26 @@ object Build {
...
+        if (args.isEmpty) {
+          println("Couldn't run `dotr` without args. Use `repl` to run the repl or add args to run the dotty application")
+        } else if (java == "") {
           println("Couldn't find java executable on path, please install java to a default location")
-        else if (scalaLib == "") {
+        } else if (scalaLib == "") {
           println("Couldn't find scala-library on classpath, please run using script in bin dir instead")
         } else {
           val dottyLib = packageAll.value("dotty-library")
           s"""$java -classpath .:$dottyLib:$scalaLib ${args.mkString(" ")}""".!
```
Comment (allanrenucci) [inline project/Build.scala]: Since you're refactoring. Triple quotes is not needed here

File: .drone.yml [hunk 4]
```diff
@@ -25,7 +25,7 @@ pipeline:
     image: lampepfl/dotty:2017-10-20
     commands:
       - cp -R . /tmp/1/ && cd /tmp/1/
-      - ./project/scripts/sbt ";compile ;testAll ;dotty-bench/jmh:run 1 1 tests/pos/alias.scala"
+      - ./project/scripts/sbt ";compile ;testAll ;dotty-bench/jmh:run 1 1 tests/pos/alias.scala; ;dotc tests/run/arraycopy.scala ;dotr Test"
```
Comment (allanrenucci) [inline .drone.yml]: `tests/pos/alias.scala; ;dotc` double semicolon?

Comment (allanrenucci) [reply .drone.yml]: Also not sure it is the best place to test an sbt task. @smarter Do you think it should be a scripted test?

File: project/Build.scala (hunk 1 above)
Comment (allanrenucci) [reply project/Build.scala]: I don't think we should special case for empty `args`. We need a better solution:
If I run `dotr -Xprint:typer`, `args` is not empty, but I still want to run the REPL

File: .drone.yml (hunk 4 above)
Comment (smarter) [reply .drone.yml]: Scripted tests are always good in my opinion :)

File: project/Build.scala (hunk 1 above)
Comment (nicolasstucki) [reply project/Build.scala]: you can at least run `repl -Xprint:typer`.

**Comment by nicolasstucki** [comment]
Next step will be to remove the `dotty.tools.dot.FromTasty` driver and integrate it to `dotty.tools.dot.Main`. This will make `-tasty` an actual `ScalaSetings` flag.

File: dist/bin/dotc [hunk 5]
```diff
@@ -82,7 +81,7 @@ case "$1" in
     # Optimize for short-running applications, see https://github.com/lampepfl/dotty/issues/222
     -Oshort) addJava "-XX:+TieredCompilation -XX:TieredStopAtLevel=1" && shift ;;
         -repl) PROG_NAME="$ReplMain" && shift ;;
-       -tasty) addScala "-Yretain-trees" && PROG_NAME="$FromTasty" && shift ;;
+       -tasty) addScala "-Yretain-trees" && PROG_NAME="$CompilerMain" && shift ;;
```
Comment (smarter) [inline dist/bin/dotc]: I would just make `-Yretain-trees` default to true when `-tasty` is used.

Comment (smarter) [reply dist/bin/dotc]: I mean, in the compiler itself, instead of doing it in various scripts.

**Comment by nicolasstucki** [comment]
Last commit needs review

File: project/scripts/sbtBootstrappedTests [hunk 6]
```diff
@@ -2,3 +2,34 @@
...
 ./project/scripts/sbt "dotty-bench-bootstrapped/jmh:run 1 1 tests/pos/alias.scala"
+
+
+# setup for `dotc`/`dotr` script tests
+./project/scripts/sbt dist-bootstrapped/pack
+
+# check that `dotc` compiles and `dotr` runs it
+echo "testing sbt dotc and dotr"
+mkdir out/scriptedtest0
+./dist-bootstrapped/target/pack/bin/dotc tests/pos/sbtDotrTest.scala -d out/scriptedtest0
```
Comment (OlivierBlanvillain) [inline project/scripts/sbtBootstrappedTests]: Any reason to use `dist-bootstrapped/**/dotc` instead of `bin/dotc`?

Comment (nicolasstucki) [reply project/scripts/sbtBootstrappedTests]: I wanted to make sure I was testing that version of the script. We could also add some tests for `bin/dotc` in the future.

Comment (OlivierBlanvillain) [reply project/scripts/sbtBootstrappedTests]: `bin/dotc` just forwards to `dist-bootstrapped/target/pack/bin/dotc` and call `sbt dist-bootstrapped/pack` if needed :)

Comment (nicolasstucki) [reply project/scripts/sbtBootstrappedTests]: Ok, I'll change it.

**Pull request merged**